pre_commit_hooks/trailing_whitespace_fixer.py

```diff
@@ -3,17 +3,33 @@
 import argparse
 import fileinput
 import sys
+import os
 
 from pre_commit_hooks.util import cmd_output
 
 
-def _fix_file(filename):
+def _fix_file(filename, markdown=False):
     for line in fileinput.input([filename], inplace=True):
+        # preserve trailing two-space for non-blank lines in markdown files
+        if markdown and (not line.isspace()) and (line.endswith("  \n")):
+            line = line.rstrip(' \n')
+            # only preserve if there are no trailing tabs or unusual whitespace
+            if not line[-1].isspace():
+                print(line + "  ")
+                continue
+
         print(line.rstrip())
 
 
 def fix_trailing_whitespace(argv=None):
     parser = argparse.ArgumentParser()
+    parser.add_argument('--no-markdown-linebreak-ext', action='store_const',
+                        const=[], default=argparse.SUPPRESS,
+                        dest='markdown_linebreak_ext',
+                        help='Do not preserve linebreak spaces in Markdown')
+    parser.add_argument('--markdown-linebreak-ext', action='append', const='*',
+                        default=argparse.SUPPRESS, metavar='EXTS', nargs='?',
+                        help='Markdown extensions (or *) for linebreak spaces')
```

Comment:
I really think the best default is for the common extensions (`.md` and `.markdown`) to get the special treatment.  While this is a bit more of an edge case that replacing tabs with spaces in a Makefile, I still think it is better to come down on the side of safety (i.e. no change to semantics) even if it means a few non-markdown `.md` files might have extraneous whitespace left in them, since at the end of the day that is a purely aesthetic consideration.  I don't expect many (any?) people to actually set these options, so the default settings are for me the most important.

Having a non-empty default requires a `store` action to clear the set.  I used an `append` action to support somebody who adds extensions one at a time, but my original code was not consistent as the first option overwrote the defaults rather than appending to them; that is too much magic. It is simpler, clearer, and safer to always append, so I made [that change instead](https://github.com/dupuy/pre-commit-hooks/commit/cc29df5d4544955713f58f5b71fc895ec7d7ca73).

The other reason for a separate `--no-markdown-linebreak-ext` option with a `store` action (rather than using a single option with a `store` action to set (or if empty, clear) the extension list) is that argparse will silently "eat" a following non-option argument (interpreting a staged filename as an extension) if you omit the `=` (e.g. `--markdown-linebreak-ext` instead of `--markdown-linebreak-ext=`).  I actually ran into that during testing, thus the `--` argument and comment in `test_markdown_linebreak_ext_opt_all`.  It's better to have the code error out if the argument provided is missing (`--` or other following option), empty (`=` or explicit empty string argument), or has a non-initial `.` (probably an innocent filename) so I [made that change too](https://github.com/dupuy/pre-commit-hooks/commit/00eb809a2518b2a2e1c99f951a1d86b97c9c2239).

Probably the most common case for the special `*` extension would be to enable linebreak preservation for files without extensions (you could use `--markdown-linebreak-ext=,` but that's pretty obscure, and I doubt anybody could figure out what that does/is supposed to do without looking at the code).

Finally, in your example you wrote `--markdown-trailing-whitespace-extension`.
I feel like the `--markdown-linebreak-ext` is already too long, but if you think the name should be changed, I'm happy to do so.